homeworks/solutions/krasnor/2-ddg/src/halfedge.js

```diff
@@ -0,0 +1,634 @@
+        this.idx = idx      // Number
+    }
+
+    normal(method = 'equal-weighted') {
+        switch (method) {
+            case 'equal-weighted':
+                // TODO: compute euqally weighted normal of this vertex
+                let eqal_weighted_sum = new Vector();
+
+                for (let f of this.getAdjacentFaces()) {
+                    let normal = f.calculateFaceNormal();
+                    eqal_weighted_sum = eqal_weighted_sum.add(normal);
+                }
+                return eqal_weighted_sum.unit();
+            case 'area-weighted':
+                // TODO: compute area weighted normal of this vertex
+                let area_weighted_sum = new Vector();
+                for (let f of this.getAdjacentFaces()) {
+                    let normal = f.calculateFaceNormal();
+                    let area = f.getArea();
+                    area_weighted_sum = area_weighted_sum.add(normal.scale(area));
+                }
+                return area_weighted_sum.unit();
+            case 'angle-weighted':
+                // TODO: compute angle weighted normal of this vertex
+                let angle_weighted_sum = new Vector();
+                let outgoingHfEdges = this.getAllOutgoingHalfedges();
+                for (let f of this.getAdjacentFaces()) {
+                    //      v2
+                    //    /   \
+                    //  prv    e1
+                    //  /       \
+                    // v0--out-->v1
+                    let normal = f.calculateFaceNormal();
+                    let v0 = outgoingHfEdges.get(f.idx);
+                    let v2 = v0.prev.twin;
+                    let angle_dot = (v0.getVector().unit().dot(v2.getVector().unit()) + 1) / 2;
+
+                    // negative values?: in a tri mesh the angle can only vary between 0 and 180 degrees
+                    if (angle_dot < 0) {
+                        let a_deg = Math.acos((angle_dot) / (v0.getVector().unit().norm() * v2.getVector().unit().norm())) * (180 / Math.PI);
+                        throw "Angle greater than 180° in Tri-Mesh: " + angle_dot + "|" + a_deg + "° Face index: " + f.idx;
+                    }
+                    angle_weighted_sum = angle_weighted_sum.add(normal.scale(angle_dot));
+                }
+                return angle_weighted_sum.unit();
+            default: // undefined
+                return new Vector()
+        }
+    }
+
+    curvature(method = 'Mean') {
+        let def = this.calcMeanCurvatureNormal(); // let angle_defect = def.dot(this.normal('equal-weighted')); // using this normal produces a better result, but not 100% sure about the math.
+
+        let H = 0.5 * def.norm(); // calculate the actual Mean Curvature value
+        let K = this.calcGaussianCurvature();
+
+        let angle_defect = K; // using angle defect as mentioned in the github discussion
+        H = angle_defect > 0 ? H : -H; // adjust color;
+
+
+        let k1 = H - Math.sqrt(H * H - K);
```

Comment:
Theoretically, `H^2-K` is always non-negative. But in practice, this can fall into negative values (because of bad triangulation/discretization).
Usually, we handle the case by falling back to mean curvature, that is `H^2-K < 0 ? 0 : H^2-K`